File: leetcode/tree/fenwick/0307_range_sum_query_mutable.py
# LC 307. Range Sum Query - Mutable | Medium (Fenwick Tree / BIT)
class NumArray:
    def __init__(self, nums):
        self.n = len(nums); self.nums = [0]*self.n; self.tree = [0]*(self.n+1)
        for i, v in enumerate(nums): self.update(i, v)
    def update(self, i, val):
        diff = val - self.nums[i]; self.nums[i] = val; i += 1
        while i <= self.n: self.tree[i] += diff; i += i & (-i)
    def _prefix(self, i):
        s = 0; i += 1
        while i > 0: s += self.tree[i]; i -= i & (-i)
        return s
    def sumRange(self, left, right):
        return self._prefix(right) - (self._prefix(left-1) if left > 0 else 0)

File: leetcode/tree/fenwick/0307_range_sum_query_mutable.cpp
// LC 307. Range Sum Query - Mutable | Medium (Fenwick Tree / BIT)
#include <vector>
using namespace std;

class NumArray {
    vector<int> tree, nums; int n;
    void add(int i, int d) { for (++i; i <= n; i += i&(-i)) tree[i] += d; }
    int prefix(int i) { int s = 0; for (++i; i > 0; i -= i&(-i)) s += tree[i]; return s; }
public:
    NumArray(vector<int>& nums) : n(nums.size()), nums(nums), tree(nums.size()+1, 0) {
        for (int i = 0; i < n; i++) add(i, nums[i]);
    }
    void update(int i, int val) { add(i, val - nums[i]); nums[i] = val; }
    int sumRange(int l, int r) { return prefix(r) - (l > 0 ? prefix(l-1) : 0); }
};
